src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs

```diff
@@ -0,0 +1,392 @@
+        {
+            // The format of BLOBHEADER (or PUBLICKEYSTRUC):
+            //  BYTE   bType
+            //  BYTE   bVersion
+            //  WORD   reserved
+            //  ALG_ID aiKeyAlg
+
+            bType = br.ReadByte();         // BLOBHEADER.bType: Expected to be 0x6 (PUBLICKEYBLOB) or 0x7 (PRIVATEKEYBLOB), though there's no check for backward compat reasons.
+            bVersion = br.ReadByte();      // BLOBHEADER.bVersion: Expected to be 0x2 or 0x3, though there's no check for backward compat reasons.
+            br.BaseStream.Position += sizeof(ushort); // BLOBHEADER.wReserved
+            int algId = br.ReadInt32();    // BLOBHEADER.aiKeyAlg
+            if (algId != CALG_DSS_SIGN)
+                throw new PlatformNotSupportedException();
+        }
+
+        private static void WriteKeyBlobHeader(DSAParameters dsaParameters, BinaryWriter bw, bool isPrivate, out bool isV3)
+        {
+            // Write out the BLOBHEADER (or PUBLICKEYSTRUC).
+            isV3 = false;
+
+            // If Y is present and this is a private key, 
+            // or Y and J are present and this is a public key, this should be a v3 blob.
+            byte version = BLOBHEADER_CURRENT_BVERSION;
+            if (((dsaParameters.Y != null) && isPrivate) ||
+                ((dsaParameters.Y != null) && (dsaParameters.J != null)))
+            {
+                Debug.Assert(dsaParameters.Y.Length > 0);
+                isV3 = true;
+                version = 0x3;
+            }
+            bw.Write((byte)(isPrivate ? PRIVATEKEYBLOB : PUBLICKEYBLOB));  // BLOBHEADER.bType
+            bw.Write((byte)version);                                       // BLOBHEADER.bVersion
+            bw.Write((ushort)0);                                           // BLOBHEADER.wReserved
+            bw.Write((int)CapiHelper.CALG_DSS_SIGN);                       // BLOBHEADER.aiKeyAlg
+        }
+
+        private static void ReadDSSSeed(DSAParameters dsaParameters, BinaryReader br, bool isV3)
+        {
+            int counter = br.ReadInt32();
+
+            if (isV3)
+            {
+                if (unchecked((uint)counter != 0xFFFFFFFF))
+                {
+                    dsaParameters.Counter = counter;
+                    dsaParameters.Seed = br.ReadReversed(20);
+                    return;
+                }
+            }
+            else
+            {
+                // This compares > 0 for backwards compatibility, and not just 0xFFFFFFFF
+                if (counter > 0) // Does not include 0xFFFFFFFF which is signed value of -1
+                {
+                    dsaParameters.Counter = counter;
+                    dsaParameters.Seed = br.ReadReversed(20);
+                    return;
+                }
+            }
+
+            dsaParameters.Counter = 0;
+            dsaParameters.Seed = null;
+            br.BaseStream.Position += 20; // Advance past seed[20]
```

Comment:
I refactored it slightly to combine the first 2 cases, but left the last += 20 like that for efficiency